### PR TITLE
[REF] [Feature] [56379] Drink Detail Screen

### PR DESCRIPTION
**Pull request by hungld-1302**
## Related Tickets
- [#Task 56379: Drink Detail Screen](https://edu-redmine.sun-asterisk.vn/issues/56379)
## WHAT
- Implement Detail Screen
- Implement Favourite Screen
## Evidence (Screenshot or Video)
image
image

## Review Checklist

Category | View Point | Description | Self review | Reviewer2 (name)
--- | --- | --- | --- | ---
Conventions | Sun* coding conventions | https://sal.vn/WtYPAy |<li>- [ ] yes</li>|<li>- [ ] yes</li>  
Conventions | Kotlin coding conventions | https://kotlinlang.org/docs/coding-conventions.html |<li>- [ ] yes</li>|<li>- [ ] yes</li>  
Redmine | Sun* Redmine working process  | https://sal.vn/piPoQn |<li>- [ ] yes</li>|<li>- [ ] yes</li>  

## Notes (Optional)
*(Impacted Areas in Application(List features, api, models or services that this PR will affect))*

*(List gem, library third party add new)*

*(Other notes)*

### CODE REVIEW

File: app/src/main/java/com/sildev/thecocktail/data/source/local/DrinkDAO.kt [hunk 1]
```diff
@@ -17,8 +17,8 @@ interface DrinkDAO {
     @Query("select * from ${DrinkDatabase.FAVORITE_TABLE}")
     suspend fun getFavoriteDrinks(): List<Drink>
 
-    @Insert(onConflict = OnConflictStrategy.REPLACE)
-    suspend fun insertFavoriteDrink(drink: Drink)
+    @Query("select * from ${DrinkDatabase.FAVORITE_TABLE} where idDrink=:id")
+    suspend fun getFavouriteById(id: String): List<Drink>
```
Comment (minhnn-1279) [inline app/src/main/java/com/sildev/thecocktail/data/source/local/DrinkDAO.kt]: :v đúng ra nó trả về 1 object hoặc null thôi chứ nhể

File: app/src/main/java/com/sildev/thecocktail/ui/favourite/FavouriteFragment.kt [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+class FavouriteFragment :
+    BaseFragment<FragmentFavouriteBinding>(FragmentFavouriteBinding::inflate) {
+
+    override val viewModel: FavouriteViewModel by viewModel()
+    private val drinkAdapter = DrinkAdapter(::onItemDrinkClick)
+
+    override fun onResume() {
+        super.onResume()
+        viewModel.requestFavourite()
```
Comment (minhnn-1279) [inline app/src/main/java/com/sildev/thecocktail/ui/favourite/FavouriteFragment.kt:23]: cái này phải để view model request favorite trước khi onResume ở base nó chạy chứ nhỉ :v

File: app/src/main/res/layout/fragment_favourite.xml [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    <LinearLayout
+        android:id="@+id/layout_empty"
+        android:layout_width="match_parent"
+        android:layout_height="match_parent"
+        android:gravity="center"
+        android:orientation="vertical">
+
+        <ImageView
+            android:layout_width="@dimen/_300dp"
+            android:layout_height="@dimen/_300dp"
```
Comment (minhnn-1279) [inline app/src/main/res/layout/fragment_favourite.xml]: https://developer.android.com/develop/ui/views/layout/constraint-layout#ratio

Với kích thước ảnh lớn thì a để theo dạng dimension ratio nhé

**Review by nhanviet1 — APPROVED** [review]

**Review by hungvv-1301 — APPROVED** [review]

File: app/src/main/java/com/sildev/thecocktail/data/source/remote/DrinkDetailDateRemoteSource.kt [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+
+import com.sildev.thecocktail.data.model.DrinkDetail
+import com.sildev.thecocktail.data.source.DrinkDetailDataSource
+import com.sildev.thecocktail.utils.Constant
+
+class DrinkDetailDateRemoteSource(private val apiService: ApiService) :
+    DrinkDetailDataSource.Remote {
+
+    override suspend fun getDetailDrink(id: String): DrinkDetail =
+        apiService.getDetailDrink(id).drinks[Constant.FIRST_INDEX]
```
Comment (ghost) [inline app/src/main/java/com/sildev/thecocktail/data/source/remote/DrinkDetailDateRemoteSource.kt]:
```suggestion
        apiService.getDetailDrink(id).drinks.first()
```

File: app/src/main/java/com/sildev/thecocktail/ui/detail/DetailActivity.kt [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+                IngredientDetail(
+                    drinkDetail.strIngredient1, drinkDetail.strMeasure1
+                )
+            )
+        }
+        if (drinkDetail.strIngredient2 != null && drinkDetail.strMeasure2 != null) {
+            listIngredient.add(
+                IngredientDetail(
+                    drinkDetail.strIngredient2, drinkDetail.strMeasure2
+                )
```
Comment (ghost) [inline app/src/main/java/com/sildev/thecocktail/ui/detail/DetailActivity.kt]: anh tham khảo đoạn này thử xem: 
image

File: app/src/main/java/com/sildev/thecocktail/data/model/DrinkDetailCollapse.kt [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+    val measures: List<String?>?,
+    val dateModified: String
+) {
+
+    fun getIngredientDetails(): List<IngredientDetail> {
+        val list = mutableListOf<IngredientDetail>()
+        val length = ingredients?.size ?: 0
+        for (i in 0 until length) {
+            ingredients?.get(i)?.let {
+                if (it.isNotBlank()) {
```
Comment (vietnh-1109) [inline app/src/main/java/com/sildev/thecocktail/data/model/DrinkDetailCollapse.kt:23]: cả đống này = ingredients.filter{ it.isNotBlank()}.foreachIndexed {...}

File: app/src/main/java/com/sildev/thecocktail/ui/favourite/FavouriteViewModel.kt [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+
+class FavouriteViewModel(private val favouriteRepository: FavouriteRepository) : BaseViewModel() {
+    private val _favouriteList = MutableLiveData<List<Drink>>()
+    val favouriteList: LiveData<List<Drink>> get() = _favouriteList
+
+    init {
+        requestFavourite()
+    }
+
+    fun requestFavourite() {
```
Comment (vietnh-1109) [inline app/src/main/java/com/sildev/thecocktail/ui/favourite/FavouriteViewModel.kt:13]:
```suggestion
   private fun requestFavourite() {
```

**Pull request merged**